### ISSUE TITLE
error: definition with same mangled name when specializing a concepted template

### ISSUE THREAD

**Issue by Alcaro** [llvm/llvm-project#58142]
```c++
#include <cstdio>
#include <concepts>

template <typename T>
void fun(T);

template <>
void fun(char c)
{
    std::puts("foo<char>()");
}

template <std::integral I>
void fun(I i)
{
    std::puts("foo<std::integral>()");
}

int main()
{
    fun<char>(' ');
}
```
Compile with `-std=c++20`.

Result:
```console
<source>:14:6: error: definition with same mangled name '_Z3funIcEvT_' as another definition
<source>:8:6: note: previous definition is here

Expected: Prints foo<char>(), or at least a less cryptic error
```

https://godbolt.org/z/axvdbK1Eh

GCC and MSVC have similar bugs:
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=107151
https://developercommunity.visualstudio.com/t/Template-explicit-specializationconcept/10012835

**Comment by usx95** [comment]
Very similar to https://github.com/llvm/llvm-project/issues/48216

**Comment by erichkeane** [comment]
While a similar look, this isn't the same issue as 48216 has a fix for the friends instantiation, this one is unofortunately different.

**Comment by zygoloid** [comment]
Fixed in 4b163e343cfa54c8d55c9da73c70d58f55ea9df2. Contrary to the suggestion at the start of this issue, the program prints
```
foo<std::integral>()
```
because the `fun(char)` explicit specialization specializes the primary template, not the constrained `template<integral>` overload, because the latter is not yet declared. And overload resolution picks the `template<integral>` version because it's more constrained.
